flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteChannelStateChecker.java

```diff
@@ -0,0 +1,125 @@
+
+	private final ResultPartitionID resultPartitionId;
+
+	private final String taskNameWithSubtask;
+
+	public RemoteChannelStateChecker(ResultPartitionID resultPartitionId, String taskNameWithSubtask) {
+		this.resultPartitionId = resultPartitionId;
+		this.taskNameWithSubtask = taskNameWithSubtask;
+	}
+
+	public boolean isProducerConsumerReadyOrAbortConsumption(CheckResult checkResult) {
+		Either<ExecutionState, Throwable> result = checkResult.getProducerExecutionState();
+		if (result.isLeft() || result.right() instanceof TimeoutException) {
+			boolean isProducerConsumerReady = isProducerConsumerReady(checkResult);
+			if (isProducerConsumerReady) {
+				return true;
+			} else {
+				abortConsumptionOrIgnoreCheckResult(checkResult);
+			}
+		} else {
+			handleFailedCheckResult(checkResult);
+		}
+		return false;
+	}
+
+	private boolean isProducerConsumerReady(CheckResult checkResult) {
+		ExecutionState consumerState = checkResult.getConsumerExecutionState();
+		Either<ExecutionState, Throwable> result = checkResult.getProducerExecutionState();
+		ExecutionState producerState = result.isLeft() ? result.left() : ExecutionState.RUNNING;
+		return consumerState == ExecutionState.RUNNING &&
+			(producerState == ExecutionState.SCHEDULED ||
+				producerState == ExecutionState.DEPLOYING ||
+				producerState == ExecutionState.RUNNING ||
+				producerState == ExecutionState.FINISHED);
+	}
+
+	private void abortConsumptionOrIgnoreCheckResult(CheckResult checkResult) {
+		ExecutionState consumerState = checkResult.getConsumerExecutionState();
+		Either<ExecutionState, Throwable> result = checkResult.getProducerExecutionState();
+		ExecutionState producerState = result.isLeft() ? result.left() : ExecutionState.RUNNING;
+		if (consumerState == ExecutionState.RUNNING) {
+			if (producerState == ExecutionState.CANCELING ||
+				producerState == ExecutionState.CANCELED ||
+				producerState == ExecutionState.FAILED) {
+
+				// The producing execution has been canceled or failed. We
+				// don't need to re-trigger the request since it cannot
+				// succeed.
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Cancelling task {} after the producer of partition {} with attempt ID {} has entered state {}.",
+						taskNameWithSubtask,
+						resultPartitionId.getPartitionId(),
+						resultPartitionId.getProducerId(),
+						producerState);
+				}
+
+				checkResult.cancelConsumption();
+			} else {
+				// Any other execution state is unexpected. Currently, only
+				// state CREATED is left out of the checked states. If we
+				// see a producer in this state, something went wrong with
+				// scheduling in topological order.
+				final String msg = String.format("Producer with attempt ID %s of partition %s in unexpected state %s.",
+					resultPartitionId.getProducerId(),
+					resultPartitionId.getPartitionId(),
+					producerState);
+
+				checkResult.failConsumption(new IllegalStateException(msg));
+			}
+		} else {
```

Comment:
The else branch would not be needed if we check the consumer state before calling into the `RemoteChannelStateChecker`.